docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md

```diff
@@ -0,0 +1,615 @@
+
+[!code-csharp[Don't report diagnostics on const declarations](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#AlreadyConst "Test to ensure no diagnostics are reported on already const definition")]
+
+Run all the unit tests now. This new test passes. After all, your analyzer doesn't do anything yet. Add the following code to `AnalyzeNode` to ensure this test continues to pass:
+
+```csharp
+// Only consider local variable declarations that aren't already const.
+if (localDeclaration.Modifiers.Any(SyntaxKind.ConstKeyword))
+{
+    return;
+}
+```
+
+### Examine declaration initializers
+
+Constant declarations are not the only case where your analyzer can exit quickly. Any variable declaration that doesn't contain an initializer cannot be made constant. Also, any variable declaration whose initializer is not a compile-time constant cannot be made const. Write two new tests that verify these two rules:
+
+[!code-csharp[Don't report diagnostics on declarations that can't be const](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#CantBeConst "Test to ensure no diagnostics are reported when the declaration can't be made const")]
+
+There's one additional C# spec rule that you should take care of with the same code: If a declaration statement includes multiple variables, they must all be const, or none of them can be const. That means one other test to ensure that a diagnostic is not reported when a declaration statement has multiple variables, and not all of them can be made const. That suggests one more test:
+
+[!code-csharp[Don't report diagnostics on multiple declarations if all can't be const](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#MultipleDeclarations "Test to ensure no diagnostics are reported when multiple declarations can't all be made const")]
+
+As before, these tests will pass now. It's time to write the code that checks for these conditions. You'll perform some semantic analysis using the <xref:Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext>. You use the `context` argument to determine whether the local variable declaration can be made `const`. A <xref:Microsoft.CodeAnalysis.SemanticModel?displayProperty=nameWithType> represents of all semantic information in a single source file. You can earn more in the article that covers [semantic models](../work-with-semantics.md). Add the following code to the `AnalyzeNode` method:
+
+```csharp
+// Ensure that all variables in the local declaration have initializers that
+// are assigned with constant values.
+foreach (var variable in localDeclaration.Declaration.Variables)
+{
+    var initializer = variable.Initializer;
+    if (initializer == null)
+    {
+        return;
+    }
+
+    var constantValue = context.SemanticModel.GetConstantValue(initializer.Value);
+    if (!constantValue.HasValue)
+    {
+        return;
+    }
+}
+```
```

Comment:
If you have a declaration like `int i = 42, j;`, this will check `GetConstantValue().HasValue` on the `42` expression, which is an expensive semantic check, even though the cheap syntactic check for whether `j` has initializer will be `false`. This could be solved by splitting the loop into two: one checks `Initializer == null`, the other `GetConstantValue().HasValue`.

Is that kind of performance optimization appropriate here, or is simplicity more important? (In which case, this code would be fine.)

Reply:
I think keeping the explanation more clear works best here. Especially because the vast majority of variable declarations declare one variable.